src/targets/gpu/fuse_special_ops.cpp

```diff
@@ -0,0 +1,194 @@
+                }
+            }
+        }
+        return inss;
+    }
+
+    void apply(module_pass_manager& mpm, const match::matcher_result& r) const
+    {
+        auto gemm2       = r.result;
+        auto gemm1       = r.instructions["dot1"];
+
+        // Capture all instructions part of the attention op
+        auto attn_inss = get_attn_instructions(gemm1, gemm2);
+
+        // Transform unordered set to sorted vector to preserve original order
+        std::vector<instruction_ref> attn_ins_vec;
+        attn_ins_vec.assign(attn_inss.begin(), attn_inss.end());
+        std::sort(
+            attn_ins_vec.begin(), attn_ins_vec.end(), [&](instruction_ref x, instruction_ref y) {
+                return std::distance(mpm.get_module().begin(), x) <
+                       std::distance(mpm.get_module().begin(), y);
+            });
+
+        // Add captured instructions to new submodule
+        module m_attn;
+        std::unordered_map<instruction_ref, instruction_ref> map_mm_to_mattn;
+        auto attn_outs = m_attn.fuse(attn_ins_vec, &map_mm_to_mattn);
+
+        // Define outputs based on instructions that are used elsewhere in the graph
+        std::vector<instruction_ref> required_outputs;
+        std::copy_if(attn_ins_vec.begin(),
+                     attn_ins_vec.end(),
+                     std::back_inserter(required_outputs),
+                     [&](auto i) {
+                         return not std::all_of(i->outputs().begin(),
+                                                i->outputs().end(),
+                                                [&](auto o) { return contains(attn_ins_vec, o); });
+                     });
+
+        assert(required_outputs.size() > 0);
+        // Not supporting multi-out just yet - TODO: remove for lse support
+        if(required_outputs.size() > 1)
+            return;
+
+        // Find corresponding output instructions in m_attn
+        std::vector<instruction_ref> m_attn_outputs;
+        std::transform(required_outputs.begin(),
+                       required_outputs.end(),
+                       std::back_inserter(m_attn_outputs),
+                       [&](auto i) { return map_mm_to_mattn.at(i); });
+
+        m_attn.add_return(m_attn_outputs);
+
+        // Define inputs to m_attn
+        auto map_mattn_to_mm = invert_map_ins(map_mm_to_mattn);
+        auto new_inputs      = m_attn.get_inputs(map_mattn_to_mm);
+
+        module_ref mpm_attn = mpm.create_module("attn" + get_count(), std::move(m_attn));
+        mpm_attn->set_bypass();
+
+        // Construct group op with the attention module
+        mpm.get_module().replace_instruction(required_outputs.front(),
+                                             make_op("group", {{"tag", "attention"}}),
+                                             new_inputs,
+                                             {mpm_attn});
+    }
+};
+
+} // namespace
+
+void fuse_special_ops::apply(module_pass_manager& mpm) const
```

Comment:
I would name this pass `fuse_attention`. It should go under `src/` and not `src/targets/gpu`.